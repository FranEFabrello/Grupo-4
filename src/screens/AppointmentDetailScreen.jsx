import React, { useState } from 'react'; // Import useState
import { View, Text, TouchableOpacity, ScrollView, Alert, ActivityIndicator } from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import { useAppTheme } from '~/providers/ThemeProvider';
import Icon from 'react-native-vector-icons/FontAwesome5';
import AppContainer from '../components/AppContainer';
import { rescheduleAppointment, cancelAppointment, fetchAppointments } from "~/store/slices/appointmentsSlice";
import { fetchNotificaciones } from '~/store/slices/notificationSlice';
import { useTranslation } from "react-i18next";
import LoadingOverlay from '~/components/LoadingOverlay'; // Import your LoadingOverlay component
import '../i18n';

function parseLocalDate(fechaStr) {
  if (!fechaStr) return null;
  if (fechaStr.includes('T')) return new Date(fechaStr);
  const [year, month, day] = fechaStr.split('-').map(Number);
  return new Date(year, month - 1, day);
}

export default function AppointmentDetailScreen({ route, navigation }) {
  const dispatch = useDispatch();
  const { colorScheme } = useAppTheme();
  const { status, error, appointmentsByUser } = useSelector((state) => state.appointments);
  const usuarioId = useSelector((state) => state.user.usuario?.id);
  const { t, i18n } = useTranslation();
  const [isCancelLoading, setIsCancelLoading] = useState(false); // Add local loading state

  const containerClass = colorScheme === 'light' ? 'bg-white' : 'bg-gray-800';
  const textClass = colorScheme === 'light' ? 'text-gray-800' : 'text-gray-200';
  const labelClass = colorScheme === 'light' ? 'text-blue-600' : 'text-blue-400';
  const cardClass = colorScheme === 'light' ? 'bg-gray-50' : 'bg-gray-700';
  const borderClass = colorScheme === 'light' ? 'border-gray-100' : 'border-gray-600';

  const notificacion = route.params?.notificacion;
  const appointmentParam = notificacion?.turno || route.params?.appointment;

  // Buscar el turno actualizado en el store por id
  //const appointmentFromStore = useSelector((state) => state.appointments.appointment);
  const appointment = appointmentsByUser.find(a => a.id === appointmentParam?.id);
  if (!appointment) {
    return <LoadingOverlay />; // o algún mensaje
  }


  const handleReschedule = () => {
    navigation.navigate('BookAppointment', {
      reprogramming: true,
      appointmentId: appointment.id,
      professionalId: appointment.doctorInfo?.id,
      specialtyId: appointment.especialidadInfo?.id,
      currentDate: appointment.fecha,
      currentStart: appointment.horaInicio,
      currentEnd: appointment.horaFin,
    });
  };

  const getStatusConfig = (estado) => {
    switch (estado) {
      case 'CONFIRMADO':
        return {
          icon: 'check-circle',
          bgColor: colorScheme === 'light' ? 'bg-green-100' : 'bg-green-900',
          textColor: colorScheme === 'light' ? 'text-green-800' : 'text-green-200',
          borderColor: colorScheme === 'light' ? 'border-green-200' : 'border-green-700',
          label: t('appointments.type.approved')
        };
      case 'CANCELADO':
        return {
          icon: 'times-circle',
          bgColor: colorScheme === 'light' ? 'bg-red-100' : 'bg-red-900',
          textColor: colorScheme === 'light' ? 'text-red-800' : 'text-red-200',
          borderColor: colorScheme === 'light' ? 'border-red-200' : 'border-red-700',
          label: t('appointments.type.cancelled')
        };
      default:
        return {
          icon: 'exclamation-circle',
          bgColor: colorScheme === 'light' ? 'bg-gray-100' : 'bg-gray-900',
          textColor: colorScheme === 'light' ? 'text-gray-800' : 'text-gray-200',
          borderColor: colorScheme === 'light' ? 'border-gray-200' : 'border-gray-700',
          label: estado
        };
    }
  };

  const statusConfig = getStatusConfig(appointment.estado);

  const handleCancel = () => {
    setIsCancelLoading(true); // Mostrar overlay

    dispatch(cancelAppointment(appointment.id))
      .unwrap()
      .then(async () => {
        await dispatch(fetchAppointmentById(appointment.id));
        setIsCancelLoading(false);
        navigation.navigate('AppointmentsScreen', { cancelled: true });
      })
      .catch((err) => {
        setIsCancelLoading(false);
        //Alert.alert('Error', typeof err === 'string' ? err : (err?.message || t('appointments.alerts.cancel_error')));
      });
  };


  return (
    <AppContainer navigation={navigation} screenTitle={t('appointments.details')}>
      <ScrollView className={`flex-1 ${containerClass}`}>
        <View className="p-5">
          {/* Cabecera */}
          <View className={`mb-6 ${cardClass} rounded-xl p-4`}>
            <View className="flex-row justify-between items-center">
              <Text className={`text-2xl font-bold ${textClass}`}>
                {t('appointments.medic_book')}
              </Text>
              <View className={`flex-row items-center ${statusConfig.bgColor} border ${statusConfig.borderColor} px-3 py-2 rounded-full`}>
                <Icon
                  name={statusConfig.icon}
                  size={16}
                  color={colorScheme === 'light'
                    ? (appointment.estado === 'CONFIRMADO' ? '#065f46' : '#991b1b')
                    : (appointment.estado === 'CONFIRMADO' ? '#6ee7b7' : '#fca5a5')}
                />
                <Text className={`ml-2 font-medium ${statusConfig.textColor}`}>
                  {statusConfig.label}
                </Text>
              </View>
            </View>
          </View>

          {/* Sección de fecha y hora */}
          <View className={`mb-4 rounded-xl p-4 ${cardClass}`}>
            <View className="flex-row items-center mb-3">
              <Icon name="calendar-alt" size={18} color={colorScheme === 'light' ? '#2563EB' : '#60A5FA'} />
              <Text className={`ml-2 ${labelClass}`}>{t('appointments.info.date_time')}</Text>
            </View>
            <Text className={`text-lg font-medium ${textClass}`}>
              {parseLocalDate(appointment.fecha).toLocaleDateString(i18n.language, {
                weekday: 'long',
                day: 'numeric',
                month: 'long',
                year: 'numeric',
              })}
            </Text>
            <Text className={`text-lg font-medium ${textClass}`}>
              {`${appointment.horaInicio} - ${appointment.horaFin}`}
            </Text>
          </View>

          {/* Sección del profesional */}
          <View className={`mb-4 rounded-xl p-4 ${cardClass}`}>
            <View className="flex-row items-center mb-3">
              <Icon name="user-md" size={18} color={colorScheme === 'light' ? '#2563EB' : '#60A5FA'} />
              <Text className={`ml-2 ${labelClass}`}>{t('appointments.info.professional')}</Text>
            </View>
            <Text className={`text-lg font-medium ${textClass}`}>
              {appointment.doctorInfo
                ? `Dr. ${appointment.doctorInfo.nombre} ${appointment.doctorInfo.apellido}`
                : t('appointments.alerts.no_assign')}
            </Text>
            <Text className={`text-base ${textClass}`}>
              {appointment.especialidadInfo?.descripcion || t('professionals.alerts.no_especiality')}
            </Text>
          </View>

          {/* Motivo de la consulta */}
          <View className={`mb-6 rounded-xl p-4 ${cardClass}`}>
            <View className="flex-row items-center mb-3">
              <Icon name="comment-medical" size={18} color={colorScheme === 'light' ? '#2563EB' : '#60A5FA'} />
              <Text className={`ml-2 ${labelClass}`}>{t('medical_note.reason')}</Text>
            </View>
            <Text className={`text-base ${textClass}`}>
              {appointment.nota || t('medical_note.no_reason')}
            </Text>
          </View>

          {/* Botón de cancelar y reprogramar */}
          {status === 'loading' || isCancelLoading ? (
            <LoadingOverlay /> // Show LoadingOverlay when canceling
          ) : appointment.estado === 'CONFIRMADO' ? (
            <View className="flex-row">
              <TouchableOpacity
                className="flex-1 bg-red-600 rounded-xl p-4 flex-row justify-center items-center shadow-sm mr-2"
                onPress={handleCancel}
              >
                <Icon name="times-circle" size={20} color="white" />
                <Text className="text-white text-base font-medium ml-2">
                  {t('global.button.cancel')}
                </Text>
              </TouchableOpacity>
              <TouchableOpacity
                className="flex-1 bg-blue-600 rounded-xl p-4 flex-row justify-center items-center shadow-sm"
                onPress={handleReschedule}
              >
                <Icon name="calendar-plus" size={20} color="white" />
                <Text className="text-white text-base font-medium ml-2">
                  {t('appointments.reschedule')}
                </Text>
              </TouchableOpacity>
            </View>
          ) : null}
        </View>
      </ScrollView>
      {isCancelLoading && <LoadingOverlay />}
    </AppContainer>
  );
}

