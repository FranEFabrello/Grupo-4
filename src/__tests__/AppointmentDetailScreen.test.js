import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import AppointmentDetailScreen from '~/screens/AppointmentDetailScreen';
import { useDispatch, useSelector } from 'react-redux';

// Mock Redux hooks
jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

// Mock navigation and route
const mockNavigate = jest.fn();
const route = {
  params: {
    appointment: {
      id: 1,
      fecha: '2025-07-10',
      horaInicio: '09:00',
      horaFin: '10:00',
      estado: 'CONFIRMADO',
      doctorInfo: { nombre: 'John', apellido: 'Doe' },
      especialidadInfo: { descripcion: 'Cardio' },
      nota: 'Consultation reason',
    },
  },
};
const navigation = { navigate: mockNavigate };

// Mock other dependencies
jest.mock('react-native-vector-icons/FontAwesome5', () => 'Icon');
jest.mock('expo-blur', () => ({ __esModule: true, BlurView: () => null }));
jest.mock('../components/LoadingOverlay', () => () => null);
jest.mock('../components/AppFooter', () => () => null);
jest.mock('react-i18next', () => ({ useTranslation: () => ({ t: key => key, i18n: { language: 'en' } }) }));
jest.mock('~/providers/ThemeProvider', () => ({ useAppTheme: () => ({ colorScheme: 'light' }) }));

describe('AppointmentDetailScreen', () => {
  beforeEach(() => {
    useDispatch.mockReturnValue(jest.fn());
    useSelector.mockImplementation(selector => selector({
      appointments: {
        status: 'idle',
        error: null,
        appointmentsByUser: [route.params.appointment],
      },
      user: {
        usuario: { id: 1 },
      },
      notifications: {
        unreadCount: 0,
      },
    }));
  });

  it('renders all appointment details correctly and matches snapshot', () => {
    const { toJSON, getByText } = render(
      <AppointmentDetailScreen route={route} navigation={navigation} />
    );
    // Header title
    expect(getByText('appointments.details')).toBeTruthy();
    // Appointment summary
    expect(getByText('appointments.medic_book')).toBeTruthy();
    expect(getByText('appointments.type.approved')).toBeTruthy();
    // Date & Time
    expect(getByText('Thursday, July 10, 2025')).toBeTruthy();
    expect(getByText('09:00 - 10:00')).toBeTruthy();
    // Professional info
    expect(getByText('Dr. John Doe')).toBeTruthy();
    expect(getByText('Cardio')).toBeTruthy();
    // Medical note reason
    expect(getByText('Consultation reason')).toBeTruthy();
    // Action buttons
    expect(getByText('global.button.cancel')).toBeTruthy();
    expect(getByText('appointments.reschedule')).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
  });

  it('navigates to reschedule screen on reschedule button press', () => {
    const { getByText } = render(
      <AppointmentDetailScreen route={route} navigation={navigation} />
    );
    fireEvent.press(getByText('appointments.reschedule'));
    expect(mockNavigate).toHaveBeenCalledWith(
      'BookAppointment',
      expect.objectContaining({ reprogramming: true })
    );
  });
});
