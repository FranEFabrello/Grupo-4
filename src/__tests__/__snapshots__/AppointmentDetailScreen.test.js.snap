// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`AppointmentDetailScreen renders all appointment details correctly and matches snapshot 1`] = `
<RCTSafeAreaView
  className="flex-1 bg-lightBackground"
>
  <View
    style={
      {
        "backgroundColor": "#fff",
        "flex": 1,
      }
    }
  >
    <View
      className="flex-row justify-between items-center p-4 bg-lightBackground border-b-2 rounded-b-lg border-double border-blue-600 "
      style={
        {
          "paddingTop": 10,
        }
      }
    >
      <View
        className="flex-row items-center"
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <Icon
            color="#6c757d"
            name="arrow-left"
            size={20}
          />
        </View>
        <View
          className="w-6 h-6 bg-blue-600 rounded-md justify-center items-center"
        >
          <Text
            className="text-white text-xs font-bold"
          >
            MB
          </Text>
        </View>
        <Text
          className="ml-2 text-lg font-semibold text-lightText"
        >
          appointments.details
        </Text>
      </View>
      <View
        className="flex-row"
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <Icon
            color="#6c757d"
            name="bell"
            size={22}
          />
        </View>
      </View>
    </View>
    <View
      className="flex-1"
    >
      <RCTScrollView
        className="flex-1 bg-white"
      >
        <View>
          <View
            className="p-5"
          >
            <View
              className="mb-6 bg-gray-50 rounded-xl p-4"
            >
              <View
                className="flex-row justify-between items-center"
              >
                <Text
                  className="text-2xl font-bold text-gray-800"
                >
                  appointments.medic_book
                </Text>
                <View
                  className="flex-row items-center bg-green-100 border border-green-200 px-3 py-2 rounded-full"
                >
                  <Icon
                    color="#065f46"
                    name="check-circle"
                    size={16}
                  />
                  <Text
                    className="ml-2 font-medium text-green-800"
                  >
                    appointments.type.approved
                  </Text>
                </View>
              </View>
            </View>
            <View
              className="mb-4 rounded-xl p-4 bg-gray-50"
            >
              <View
                className="flex-row items-center mb-3"
              >
                <Icon
                  color="#2563EB"
                  name="calendar-alt"
                  size={18}
                />
                <Text
                  className="ml-2 text-blue-600"
                >
                  appointments.info.date_time
                </Text>
              </View>
              <Text
                className="text-lg font-medium text-gray-800"
              >
                Thursday, July 10, 2025
              </Text>
              <Text
                className="text-lg font-medium text-gray-800"
              >
                09:00 - 10:00
              </Text>
            </View>
            <View
              className="mb-4 rounded-xl p-4 bg-gray-50"
            >
              <View
                className="flex-row items-center mb-3"
              >
                <Icon
                  color="#2563EB"
                  name="user-md"
                  size={18}
                />
                <Text
                  className="ml-2 text-blue-600"
                >
                  appointments.info.professional
                </Text>
              </View>
              <Text
                className="text-lg font-medium text-gray-800"
              >
                Dr. John Doe
              </Text>
              <Text
                className="text-base text-gray-800"
              >
                Cardio
              </Text>
            </View>
            <View
              className="mb-6 rounded-xl p-4 bg-gray-50"
            >
              <View
                className="flex-row items-center mb-3"
              >
                <Icon
                  color="#2563EB"
                  name="comment-medical"
                  size={18}
                />
                <Text
                  className="ml-2 text-blue-600"
                >
                  medical_note.reason
                </Text>
              </View>
              <Text
                className="text-base text-gray-800"
              >
                Consultation reason
              </Text>
            </View>
            <View
              className="flex-row"
            >
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "opacity": 1,
                  }
                }
              >
                <Icon
                  color="white"
                  name="times-circle"
                  size={20}
                />
                <Text
                  className="text-white text-base font-medium ml-2"
                >
                  global.button.cancel
                </Text>
              </View>
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "opacity": 1,
                  }
                }
              >
                <Icon
                  color="white"
                  name="calendar-plus"
                  size={20}
                />
                <Text
                  className="text-white text-base font-medium ml-2"
                >
                  appointments.reschedule
                </Text>
              </View>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </View>
</RCTSafeAreaView>
`;
